File: Sources/CountriesFlagsObjC/source/Countries/UIImage+NorthMacedonia.m
//
//  UIImage+NorthMacedonia.m
//  CountriesFlags
//

#import "Countries/UIImage+NorthMacedonia.h"

#import "Extensions/CGSize+Extension.h"

@implementation UIImage (NorthMacedonia)

+ (UIImage * _Nullable)flagNorthMacedoniaWithSize:(CGSize)size {
    UIColor* colorYellow = [UIColor colorWithRed:249.0f/255.0f
                                           green:214.0f/255.0f
                                            blue:22.0f/255.0f
                                           alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:206.0f/255.0f
                                        green:32.0f/255.0f
                                         blue:40.0f/255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, NO, 0.0f);
    
    [colorRed setFill];
    [colorRed setStroke];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorYellow setFill];
    // left top
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointZero];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 0.3f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.1f, size.height * 0.56f)];
    [path closePath];
    [path fill];

    // left bottom
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 0.3f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.1f, size.height * 0.44f)];
    [path closePath];
    [path fill];

    // right bottom
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.7f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 0.9f, size.height * 0.44f)];
    [path closePath];
    [path fill];

    // right top
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.7f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 0.9f, size.height * 0.56f)];
    [path closePath];
    [path fill];

    // left center
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height * 0.4f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(0.0f, size.height * 0.6f)];
    [path closePath];
    [path fill];

    // right center
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width, size.height * 0.4f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height * 0.6f)];
    [path closePath];
    [path fill];

    // top center
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 2.0f * 0.9f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 2.0f * 0.8f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.1f, 0.0f)];
    [path closePath];
    [path fill];

    // bottom center
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 2.0f * 0.9f, size.height)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height / 2.0f * 1.2f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f * 1.1f, size.height)];
    [path closePath];
    [path fill];

    path = [UIBezierPath bezierPathWithArcCenter:CGSizeMidPoint(size)
                                          radius:size.height / 3.5f / 2.0f
                                      startAngle:0.0f
                                        endAngle:M_PI * 2.0f
                                       clockwise:YES];
    path.lineWidth = size.height / 14.0f;
    [path stroke];
    [path fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

@end

File: Sources/CountriesFlags/Countries/UIImage+NorthMacedonia.swift
//
//  UIImage+NorthMacedonia.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_North_Macedonia "wikipedia"
     
     Generates the flag of North Macedonia.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of North Macedonia.
     */
    internal static func flagNorthMacedonia(with size: CGSize) -> UIImage? {
        let colorYellow = UIColor(red: 249.0 / 255.0,
                                  green: 214.0 / 255.0,
                                  blue: 22.0 / 255.0,
                                  alpha: 1.0)
        let colorRed = UIColor(red: 206.0 / 255.0,
                               green: 32.0 / 255.0,
                               blue: 40.0 / 255.0,
                               alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorRed.setFill()
        colorRed.setStroke()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorYellow.setFill()
        // left top
        var path = UIBezierPath()
        path.move(to: .zero)
        path.addLine(to: CGPoint(x: size.width / 2.0 * 0.3, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.1, y: size.height * 0.56))
        path.close()
        path.fill()

        // left bottom
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 0.3, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.1, y: size.height * 0.44))
        path.close()
        path.fill()

        // right bottom
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.7, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 0.9, y: size.height * 0.44))
        path.close()
        path.fill()

        // right top
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.7, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 0.9, y: size.height * 0.56))
        path.close()
        path.fill()

        // left center
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height * 0.4))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: 0.0, y: size.height * 0.6))
        path.close()
        path.fill()

        // right center
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width, y: size.height * 0.4))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height * 0.6))
        path.close()
        path.fill()

        // top center
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 2.0 * 0.9, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height / 2.0 * 0.8))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.1, y: 0.0))
        path.close()
        path.fill()

        // bottom center
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 2.0 * 0.9, y: size.height))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height / 2.0 * 1.2))
        path.addLine(to: CGPoint(x: size.width / 2.0 * 1.1, y: size.height))
        path.close()
        path.fill()

        path = UIBezierPath(arcCenter: size.midPoint,
                            radius: size.height / 3.5 / 2.0,
                            startAngle: 0.0,
                            endAngle: .pi * 2.0,
                            clockwise: true)
        path.lineWidth = size.height / 14.0
        path.stroke()
        path.fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
